[ronnnnnnnnn] iP

ronnnnnnnnn opened the pull request:
### BadGPT

> "Quite possibly the worst chat bot to ever exist" - Steve Jobs
> "A disaster... I don't have too many words" - James Worthy ([source](https://www.talkbasket.net/138802-james-worthy-says-lakers-blowout-loss-vs-spurs-was-a-disaster))

BadGPT tracks tasks and nothing else. Why is it called a chat bot again? Anyway, it's

- text-based
- ~bad~ good

All you have to do is:

1. Download it from [here](https://github.com/ronnnnnnnnn/ip/releases/tag/A-Jar)
2. Double-click it
3. Add your tasks
4. Let it manage your tasks for you 😉

It _is_ free, but you might as well pay for something better.

Features:

- [x] Managing tasks
- [ ] Sentience (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ronnnnnnnnn committed:
Level-0

- Renamed Duke to BadGPT
- Added greeting and exit

ronnnnnnnnn committed:
Level-1

- Added echo functionality

ronnnnnnnnn committed:
Level-2

- Line printing is now handled in a method
- Added functionality to store user input and display it in order when requested

ronnnnnnnnn committed:
Level-3

- Added Task class
- Added basic todo list functionality

ronnnnnnnnn committed:
Level-4

- Added subclasses for Task

ronnnnnnnnn committed:
A-TextUiTesting

- Added basic automatic testing using a bat file
- Fixed an oversight where a new Scanner is created every loop, which resulted in the program not being able to read multiple lines of input at once

ronnnnnnnnn committed:
Level-5

- Handled the 2 possible errors

ronnnnnnnnn committed:
Level-6

- Added delete task function

ronnnnnnnnn committed:
Add Exceptions

ronnnnnnnnn committed:
Add FileManager.java

Task list now saved between sessions

ronnnnnnnnn committed:
Add dates

Added LocalDate and rewrote some parts of FileManager to accomodate the changes

ronnnnnnnnn committed:
Add more OOP

Added a lot more classes and exceptions

ronnnnnnnnn committed:
Add packages

Separated classes into packages

ronnnnnnnnn committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

ronnnnnnnnn committed:
Update build.gradle

ronnnnnnnnn committed:
Add JUnit tests

Added unit tests for the TaskList and Parser classes

ronnnnnnnnn committed:
Update build.gradle

ronnnnnnnnn committed:
Add JavaDoc

ronnnnnnnnn committed:
Follow coding standard

ronnnnnnnnn committed:
Add find command

ronnnnnnnnn committed:
Merge branch 'branch-A-CodingStandard'

ronnnnnnnnn committed:
Merge branch 'branch-Level-9'

ronnnnnnnnn committed:
Update find command

LargeCrowd commented on src/main/java/badgpt/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Changes the status of the task to incomplete.
+     */
+    public void uncomplete() {
```
i think theres a slight spelling error here
```suggestion
    public void incomplete() {
```

martinng01 replied:
Perhaps rename to markIncomplete()? This way it is the direct opposite of markComplete().

LargeCrowd commented on src/main/java/badgpt/util/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @param task The Task object to be stored in the list.
+     * @throws ListFullException If the list already has 100 tasks.
+     */
+    public void store(Task task) throws ListFullException {
```
would it be better to give a clearer, verb form of the method name, such as storeTask()?
```suggestion
    public void storeTask(Task task) throws ListFullException {
```

LargeCrowd commented on src/main/java/badgpt/util/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @throws TaskNotFoundException If the number entered does not exist in the list.
+     * @throws SameStatusException If the task is already complete.
+     */
+    public void mark(int taskNum) throws TaskNotFoundException, SameStatusException {
```
Could have a clearer naming as to what it does (i.e. markDone)
```suggestion
    public void markDone(int taskNum) throws TaskNotFoundException, SameStatusException {
```

LargeCrowd commented on src/main/java/badgpt/util/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @throws TaskNotFoundException If the number entered does not exist in the list.
+     * @throws SameStatusException If the task is already incomplete.
+     */
+    public void unmark(int taskNum) throws TaskNotFoundException, SameStatusException {
```
Same suggestion as for markDone
```suggestion
    public void markUndone(int taskNum) throws TaskNotFoundException, SameStatusException {
```

LargeCrowd commented on src/main/java/badgpt/util/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @param taskNum The position of the task in the list.
+     * @throws TaskNotFoundException If the number entered does not exist in the list.
+     */
+    public void delete(int taskNum) throws TaskNotFoundException {
```
same as above
```suggestion
    public void deleteTask(int taskNum) throws TaskNotFoundException {
```

LargeCrowd commented on src/main/java/badgpt/util/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     *
+     * @param keyword The term of interest to filter out tasks.
+     */
+    public void find(String keyword) throws ListEmptyException {
```
```suggestion
    public void findKeyword(String keyword) throws ListEmptyException {
```

LargeCrowd reviewed (commented):
Overall easy to read and looks good, except for the small things like method naming could be clearer/more specific. Good job!

JustWeiHao commented on src/main/java/badgpt/BadGpt.java:
Nice naming

JustWeiHao commented on src/main/java/badgpt/exceptions/BadException.java:
I like the idea of having a superclass and multiple subclasses for self-defined exceptions.

JustWeiHao commented on src/main/java/badgpt/util/TasksUi.java:
I like the personality of your chatbot.

martinng01 commented on src/main/java/badgpt/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Changes the status of the task to complete.
+     */
+    public void complete() {
```
Perhaps rename to markComplete()? Makes it clear that the function is an action.

martinng01 commented on src/main/java/badgpt/util/Parser.java:
Could consider abstracting commands out into a separate class?

martinng01 reviewed (commented):
Overall clean code, just added a few renaming suggestions.

JustWeiHao commented on src/main/java/badgpt/util/FileManager.java:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @param data The data to be written.
+     */
+    public void writeToFile(String data) {
```
Perhaps the naming of the method can be consistent (either preposition is added or omitted).
You have `loadFile`, `readFile`, and `writeToFile`.

JustWeiHao commented on src/main/java/badgpt/util/TasksUi.java:
```diff
@@ -0,0 +1,72 @@
+     * @param task The task which has been deleted.
+     * @param numTasks The current number of tasks in the list.
+     */
+    public void printDelete(Task task, int numTasks) {
```
Perhaps a more consistent naming of the methods
You have `printAddTask`, `printMarkUnmarkOutcome`, and `printDelete`

JustWeiHao reviewed (commented):
LGTM. Just some nitpick in method naming.
I wrote this second review because I just realized I need to comment on the coding standard

shunjieee commented on src/main/java/badgpt/BadGpt.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * The main class for the chatbot program.
+ */
+public class BadGpt {
+    private final String NAME = "BadGPT";
+    private TaskList taskList;
+    private Ui ui;
+    private FileManager fileManager;
```
Well abstracted, there isn't any striking violation to coding standards. :)

shunjieee reviewed (commented):
LGTM

ronnnnnnnnn committed:
Add GUI

ronnnnnnnnn committed:
Update build.gradle

ronnnnnnnnn committed:
Add assertions

Adding assertions helps to ensure that assumptions about the program
state in a particular point in execution hold.

This lets us identify potential bugs in a more timely manner.

ronnnnnnnnn committed:
Improve code quality

Some comments are self-explanatory.

Removing them improves code readability.

ronnnnnnnnn committed:
Merge pull request #1 from ronnnnnnnnn/branch-A-CodeQuality

Improve code quality

ronnnnnnnnn committed:
Merge branch 'master' into branch-A-Assertions

ronnnnnnnnn committed:
Merge pull request #2 from ronnnnnnnnn/branch-A-Assertions

Add assertions

ronnnnnnnnn committed:
Add schedule

ronnnnnnnnn committed:
Add Commands

ronnnnnnnnn committed:
Update README.md

ronnnnnnnnn committed:
Update README.md

ronnnnnnnnn committed:
Update JUnit tests
